HeterogeneousCore/CUDAServices/src/CUDAService.cc

```diff
@@ -490,3 +496,19 @@ void CUDAService::free_host(void *ptr) {
     cuda::throw_if_error(cudaFreeHost(ptr));
   }
 }
+
+
+// CUDA stream cache
+struct CUDAService::CUDAStreamCache {
+  explicit CUDAStreamCache(int ndev): cache(ndev) {}
+
+  // Separate caches for each device for fast lookup
+  std::vector<edm::ReusableObjectHolder<cuda::stream_t<>>> cache;
+};
+
+std::shared_ptr<cuda::stream_t<>> CUDAService::getCUDAStream() {
+  return cudaStreamCache_->cache[getCurrentDevice()].makeOrGet([](){
+      auto current_device = cuda::device::current::get();
+      return std::make_unique<cuda::stream_t<>>(current_device.create_stream(cuda::stream::implicitly_synchronizes_with_default_stream));
```

Comment:
there is a comment in the CUDA API Wrapper sources that I did not understand, but found worrying:

> I'm a worried about the creation of streams with the assumption
> that theirs is the current device, so I'm just forbidding it
> outright here - even though it's very natural to want to write
> ```c++
> cuda::device::curent::get().create_stream()
> ```

Why don't we simply use `cudaStreamCreate()` or `cudaStreamCreateWithFlags()` ?
If we store directly the `cudaStream_t` instead of the `cuda::stream_t` it should also get rid of one level of pointer indirection.

Reply:
My main motivation for using the CUDA API Wrapper `cuda::stream_t<>` was `cuda::stream<>::enqueue::callback()` for registering the callback for lambda (ok, the same can be done of course by ourselves as well).

Removing a level of pointer indirection requires a bit more work though
* the concrete type `std::remove_pointer<cudaStream_t>::type` doesn't seem work well with smart pointers (too well hidden etc)
* without `shared`/`unique_ptr` would need a special implementation of `edm::ReusableObjectHolder` (I believe the underlying `tbb::concurrent_queue` could still be used), and an external-to-pointer mechanism to track the refcounts (e.g. `tbb::concurrent_hash_map` within `CUDAService::CUDAStreamCache`)

Maybe for a future PR (and an issue in the mean time)?

Reply:
I see.

No, I would not try to remove the usage of `shared_ptr`/`unique_ptr` (unless we find a way to handling the streams that does not require the reference counting).

Let's go with an issue until we figure it out - according to Vincenzo the current system does not work very well with multiple GPUs anyway...

Reply:
@fwyzard 
I created an issue #279. I did not include (yet) your comment about the additional pointer indirection, because we'd have that even with using `edm::ReusableObjectHolder` directly with `cudaStream_t` (`cuda::stream_t<>` holds `cudaStream_t` by value). If you want to elaborate, could you continue on that issue (assuming I got your point correctly)?